File: DataStructures-Algorithms/12. Dynamic Programming/Longest_common_subsequence_2.c
/*
Given two sequences, find the length of longest subsequence present in both of them. 
A subsequence is a sequence that appears in the same relative order, but not necessarily contiguous.
The Sequences are of length : m,n

(Dynamic Programming - Bottom Up approach)

Time Complexity : O(m*n)

Space Complexity : O(m*n)
*/

#include<stdio.h>
#include<string.h>

#define max(a,b)  (a > b) ? a : b 

int getLCS( char X[], char Y[], int m, int n)
{
	int mat[m+1][n+1];
		
	for(int i=0; i<=m; ++i)
	{
		for(int j=0; j<=n; ++j)
		{
			
			if(i==0 || j==0)
				mat[i][j] = 0;
			
			else if(X[i-1] == Y[j-1])
				mat[i][j] = 1 + mat[i-1][j-1];
			
			else
				mat[i][j] = max(mat[i-1][j], mat[i][j-1]);
		}
	}
	
	return mat[m][n];	
}

void main()
{	
	char X[] = "AGGTAB";
	char Y[] = "GXTXAYB";
	
	int m = strlen(X);
	int n = strlen(Y);
	
	printf(" Longest Common Subsequence : %d ", getLCS(X,Y,m,n));
}

File: CODING/12. Dynamic Programming/intoJava/Longest_common_subsequence_2.java
/*
Given two sequences, find the length of longest subsequence present in both of them. 
A subsequence is a sequence that appears in the same relative order, but not necessarily contiguous.
The Sequences are of length : m,n

(Dynamic Programming - Bottom Up approach)

Time Complexity : O(m*n)

Space Complexity : O(m*n)
*/

class LCS {

	int getLCS(char[] X, char[] Y, int m, int n) {

		int L[][] = new int[m+1][n+1];

		/* Following steps build L[m+1][n+1] in bottom up fashion. Note
         	that L[i][j] contains length of LCS of X[0..i-1] and Y[0..j-1] */
		for(int i = 0; i <= m; ++i) {

			for(int j = 0; j <= n; ++j) {

				if(i == 0 || j == 0 ) {
					L[i][j] = 0;
				}
				else if(X[i-1] == Y[j-1]) {
					L[i][j] = 1 + L[i-1][j-1];
				}
				else {
					L[i][j] = Math.max(L[i-1][j], L[i][j-1]);
				}
			}
		}
		return L[m][n];
	}

	public static void main(String args[]) {

		LCS obj = new LCS();

		String s1 = "AGGTAB";
		String s2 = "GXTXAYB";

		char[] X = s1.toCharArray();
	       	char[] Y = s2.toCharArray();

		int m = X.length;
		int n = Y.length;

		System.out.println("Length of the Longest Common Subsequence: " + obj.getLCS(X, Y, m, n));
	}
}
